File: src/server/updater_v2/updatemanager.cpp
/*
 * Infomaniak kDrive - Desktop
 * Copyright (C) 2023-2024 Infomaniak Network SA
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "updatemanager.h"

#include "linuxupdater.h"
#include "sparkleupdater.h"
#include "windowsupdater.h"
#include "db/parmsdb.h"
#include "log/log.h"
#include "requests/parameterscache.h"

namespace KDC {

UpdateManager::UpdateManager(QObject *parent) : QObject(parent) {
    createUpdater();

    connect(&_updateCheckTimer, &QTimer::timeout, this, &UpdateManager::slotTimerFired);

    // Note: the sparkle-updater is not an KDCUpdater
    // if (KDCUpdater *updater = qobject_cast<KDCUpdater *>(UpdaterServer::instance())) {
    //     connect(updater, &KDCUpdater::newUpdateAvailable, this, &UpdaterScheduler::updaterAnnouncement);
    //     connect(updater, &KDCUpdater::requestRestart, this, &UpdaterScheduler::requestRestart);
    // }

    // static constexpr auto checkInterval = std::chrono::hours(1);
    static constexpr auto checkTestInterval = std::chrono::seconds(15);
    _updateCheckTimer.start(std::chrono::milliseconds(checkTestInterval).count());

    // Setup callback for update state change notification
    const std::function<void(UpdateStateV2)> callback = std::bind_front(&UpdateManager::onUpdateStateChange, this);
    _updater->setStateChangeCallback(callback);
    connect(this, &UpdateManager::updateStateChanged, this, &UpdateManager::slotUpdateStateChanged, Qt::QueuedConnection);

    // At startup, do a check in any case and setup distribution channel.
    QTimer::singleShot(3000, this, [this]() { setDistributionChannel(readDistributionChannelFromDb()); });
}

void UpdateManager::startInstaller() const {
    LOG_DEBUG(Log::instance()->getLogger(), "startInstaller called!");
    _updater->startInstaller();
}

void UpdateManager::slotTimerFired() const {
    _updater->checkUpdateAvailable();
}

void UpdateManager::slotUpdateStateChanged(const UpdateStateV2 newState) {
    LOG_DEBUG(Log::instance()->getLogger(), "New update state: " << newState);
    switch (newState) {
        case UpdateStateV2::UpToDate:
        case UpdateStateV2::Checking:
        case UpdateStateV2::Downloading: {
            // Nothing to do
            break;
        }
        case UpdateStateV2::ManualUpdateAvailable: {
            emit updateAnnouncement(tr("New update available."),
                                    tr("Version %1 is available for download.").arg(_updater->versionInfo().tag.c_str()));
            break;
        }
        case UpdateStateV2::Available: {
            // A new version has been found
            _updater->onUpdateFound();
            break;
        }
        case UpdateStateV2::Ready: {
            // TODO : manage skipped version
            // The new version is ready to be installed
#if defined(__APPLE__)
            _updater->startInstaller();
#elif defined(_WIN32)
            emit showUpdateDialog();
#endif
            break;
        }
        case UpdateStateV2::DownloadError:
        case UpdateStateV2::CheckError:
        case UpdateStateV2::UpdateError: {
            // An error occurred
            break;
        }
    }
}

void UpdateManager::createUpdater() {
#if defined(__APPLE__) && defined(HAVE_SPARKLE)
    _updater = std::make_unique<SparkleUpdater>();
#elif defined(_WIN32)
    _updater = std::make_unique<WindowsUpdater>();
#else
    // the best we can do is notify about updates
    _updater = std::make_unique<LinuxUpdater>();
#endif
}

void UpdateManager::onUpdateStateChange(const UpdateStateV2 newState) {
    // Emit signal in order to run `slotUpdateStateChanged` in main thread
    emit updateStateChanged(newState);
}

DistributionChannel UpdateManager::readDistributionChannelFromDb() const {
    return ParametersCache::instance()->parameters().distributionChannel();
}

void UpdateManager::skipVersion(const std::string &skippedVersion) const {
    // TODO : write skipped version in DB
}

} // namespace KDC

File: src/gui4/windows/kDrive client/kDrive client/ViewModels/Settings/UpdateManager.cs
/*
 * Infomaniak kDrive - Desktop
 * Copyright (C) 2023-2026 Infomaniak Network SA
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
using Infomaniak.kDrive.ServerCommunication.Interfaces;
using Infomaniak.kDrive.Types;
using Microsoft.Extensions.DependencyInjection;
using System.Threading;
using System.Threading.Tasks;

namespace Infomaniak.kDrive.ViewModels
{
    public class UpdateManager : UISafeObservableObject
    {
        private bool _updateEnabled = false;
        private bool _autoUpdateEnabled = false;
        private DistributionChannel _currentChannel = DistributionChannel.Beta;
        private AppVersion? _availableUpdate;
        private bool _fetchingUpdate = false;
        private bool _showNotification = false;

        public bool UpdateEnabled
        {
            get => _updateEnabled;
            set => SetPropertyInUIThread(ref _updateEnabled, value);
        }

        public bool AutoUpdateEnabled
        {
            get => _autoUpdateEnabled;
            set => SetPropertyInUIThread(ref _autoUpdateEnabled, value);
        }

        public DistributionChannel CurrentChannel
        {
            get => _currentChannel;
            set => SetPropertyInUIThread(ref _currentChannel, value);
        }
        public AppVersion? AvailableUpdate
        {
            get => _availableUpdate;
            set
            {
                SetPropertyInUIThread(ref _availableUpdate, value);
            }
        }

        public bool FetchingUpdate
        {
            get => _fetchingUpdate;
            set => SetPropertyInUIThread(ref _fetchingUpdate, value);
        }

        public bool ShowNotification
        {
            get => _showNotification;
            set => SetPropertyInUIThread(ref _showNotification, value);
        }

        public static async Task<bool> StartUpdate()
        {
            return await App.ServiceProvider.GetRequiredService<IServerCommService>().StartUpdate(CancellationToken.None);
        }

        public async Task<bool> SkipVersion()
        {
            bool res = await App.ServiceProvider.GetRequiredService<IServerCommService>().SkipVersion(CancellationToken.None);
            if (res)
                ShowNotification = false;
            return res;
        }

        public async Task<bool> ChangeChannel(DistributionChannel newChannel)
        {
            var previousChannel = CurrentChannel;
            CurrentChannel = newChannel;

            if (!await App.ServiceProvider.GetRequiredService<IServerCommService>().SaveSettings(CancellationToken.None))
            {
                CurrentChannel = previousChannel;
                return false;
            }
            AvailableUpdate = null;
            ShowNotification = false;
            return true;
        }

        public async Task<bool> ChangeAutoUpdate(bool activated)
        {
            AutoUpdateEnabled = activated;
            if (!await App.ServiceProvider.GetRequiredService<IServerCommService>().SaveSettings(CancellationToken.None))
            {
                AutoUpdateEnabled = !activated;
                return false;
            }
            return true;
        }
    }
}
